### PR TITLE
feat(InventoryTable): Use REST API for systems

### PR DESCRIPTION
**Pull request by bastilian**
This is an early PoC of using the REST API for the Inventory Table.

Most code has been moved over from https://github.com/RedHatInsights/hooks-frontend and slightly adapted.

It implements an `useCollection` hook which should allow to load, filter and paginate REST API resources.

### CODE REVIEW

File: src/Utilities/hooks/api/utils/normalize.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import _ from 'lodash';
```
Comment (bastilian) [inline src/Utilities/hooks/api/utils/normalize.js:1]: This is basically 1:1 from hooks, in desperate need of a refactor.

File: src/Utilities/hooks/api/utils/normalize.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+export const pendingMessage = (base) => {
+    return base + '_PENDING';
+};
+
+export const failureMessage = (base) => {
+    return base + '_REJECTED';
+};
+
+export const findItemById = (id, array) =>
+    array.find((item) => parseInt(item.id) === parseInt(id));
```
Comment (akofink) [inline src/Utilities/hooks/api/utils/normalize.js:28]: `id` and `item.id` here are both UUID strings

```suggestion
    array.find((item) => item.id === id);
```

File: src/Utilities/hooks/api/utils/normalize.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            let additionalData = includeItemRelationships(findIncluded(fullPayload, id, relatedResourceKind), fullPayload);
+
+            relatedResources[relatedResourceKind][id] = {
+                ...itemData,
+                ...additionalData
+            };
+        });
+    });
+    return { ...item, ...relatedResources };
+};
```
Comment (akofink) [inline src/Utilities/hooks/api/utils/normalize.js:63]: This function is infinitely recursive.

**Review by akofink — COMMENTED** [review]
Pretty exciting! Needs some work to get it functional. Thanks, @bastilian!

File: src/Utilities/hooks/api/useCollection.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import { normalizeData } from './utils/normalize';
+import { default as useReducer} from 'Utilities/hooks/useReducerWithLogger';
+
+const FETCH_COLLECTION = 'FETCH_COLLECTION';
+
+const fetchCollection = async (apiClient, collection, params) => {
+    const json = await apiClient.get(null, {
+        params: {
+            ...params,
+            include: ['profiles']
```
[the remark this answers is not in the document]
Comment (bastilian) [reply src/Utilities/hooks/api/useCollection.js:13]: No, not at all, of course. :)

File: src/Utilities/hooks/api/utils/normalize.js (hunk 3 above)
Comment (bastilian) [reply src/Utilities/hooks/api/utils/normalize.js:63]: Yeah, I am not sure if we should use this at all. It is a 1:1 copy of some old code from hooks/notifications to accomplish including associations in a more accessible manner.

We should see if there isn't already a more robust and tested library or snippet somewhere that does the same.

**Comment by bastilian** [comment]
@akofink It's a start, it really was just trying to see how far we can get it with the API. I'll Incorporate the suggested changes and make it more functional.

**Review by akofink — COMMENTED** [review]
It does seem to fix the double refresh and missing system name issues. :tada: 

It still uses graphql, as expected from our discussion in standup, but the PR is very misleading in that regard.

All the filters are broken, as advertised.

There are some linting errors preventing build:

```
ERROR in 
/home/akofink/dev/repos/compliance-frontend/src/SmartComponents/SystemsTable/Columns.js
  6:7  error  'logArgs' is assigned a value but never used  no-unused-vars

/home/akofink/dev/repos/compliance-frontend/src/SmartComponents/SystemsTable/InventoryTable.js
   1:34  error  'useState' is defined but never used        no-unused-vars
   1:42  error  Unexpected trailing comma                   comma-dangle
  89:13  error  Identifier 'per_page' is not in camel case  camelcase

✖ 4 problems (4 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.


webpack 5.21.2 compiled with 1 error and 7 warnings in 27209 ms
```

File: src/Utilities/ApiClient.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
+class ApiClient {
```
Comment (akofink) [inline src/Utilities/ApiClient.js:1]: Hey @bastilian - could we please get this ApiClient and useCollection functionality split out so we can use it for COMP-E-133 when fetching the supported_ssgs endpoint? We don't need it to be able to resolve relationships or includes or anything, just the basic response body.

**Comment by bastilian** [comment]
Closing this, since REST API parts of it were introduced via COMP-E-133 and others will follow with refactoring and cleanup of the InventoryTable.

**Pull request closed without merging**